File: vs2003_cd01/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/Samples/Tutorials/clr_fundamentals/samples/threading/asynchronous/vb/Asynchronous.vb
Imports System
Imports System.Threading


Class App
   Shared Sub Main()
      Console.WriteLine("Main thread: Queuing an asynchronous operation.")
      Dim asyncOpIsDone As New AutoResetEvent(False)
      ThreadPool.QueueUserWorkItem(New WaitCallback(AddressOf MyAsyncOperation), asyncOpIsDone)

      Console.WriteLine("Main thread: Performing other operations.")
      ' ...

      Console.WriteLine("Main thread: Waiting for asynchronous operation to complete.")
      asyncOpIsDone.WaitOne()

      Console.Write("Press Enter to close window...")
      Console.Read()
   End Sub


   ' The callback method's signature MUST match that of a System.Threading.WaitCallback 
   ' delegate (it takes an Object parameter and returns void)
   Shared Sub MyAsyncOperation(state As Object)
      Console.WriteLine("WorkItem thread: Performing asynchronous operation.")
      ' ...
      Thread.Sleep(5000)	' Sleep for 5 seconds to simulate doing work

      ' Signal that the async operation is now complete.
      CType(state, AutoResetEvent).Set()
   End Sub
End Class

File: vs2003_cd01/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/Samples/Tutorials/clr_fundamentals/samples/threading/asynchronous/cs/Asynchronous.cs
using System;
using System.Threading;


class App {
   public static void Main() {
      Console.WriteLine("Main thread: Queuing an asynchronous operation.");
      AutoResetEvent asyncOpIsDone = new AutoResetEvent(false);
      ThreadPool.QueueUserWorkItem(new WaitCallback(MyAsyncOperation), asyncOpIsDone);

      Console.WriteLine("Main thread: Performing other operations.");
      // ...

      Console.WriteLine("Main thread: Waiting for asynchronous operation to complete.");
      asyncOpIsDone.WaitOne();

      Console.Write("Press Enter to close window...");
      Console.Read();
   }


   // The callback method's signature MUST match that of a System.Threading.WaitCallback 
   // delegate (it takes an Object parameter and returns void)
   static void MyAsyncOperation(Object state) {
      Console.WriteLine("WorkItem thread: Performing asynchronous operation.");
      // ...
      Thread.Sleep(5000);	// Sleep for 5 seconds to simulate doing work

      // Signal that the async operation is now complete.
      ((AutoResetEvent)state).Set();
   }
}